### ISSUE TITLE
Corrupt cache leads to invalid_grant

### ISSUE THREAD

**Issue by eveliotc** [googleapis/google-api-php-client#101]
Hello,

**tl;dr**: Jump to _Finally_

Two weeks ago we migrated to 1.0.3-beta from 0.6.2, we use it mostly for authenticate calls using an auth code from our Android app. After migration in most of our servers we started getting exceptions from `verifyIdToken` with message `invalid_grant`, after some research this seems to be a pretty generic message, we made sure all of our servers had proper time even if we have always used ntpd, we made sure IPs were whitelisted as before in the proper Google API Project, made sure the codes from our Android app were properly working as this call would always return proper auth token

``` bash
curl -X POST --data-urlencode 'grant_type=authorization_code' --data-urlencode 'redirect_uri=' --data-urlencode 'client_id=ID@developer.gserviceaccount.com' --data-urlencode 'client_secret=dASeCrEt' --data-urlencode 'code=codeFromGoogleAuthUtilgetToken' https://accounts.google.com/o/oauth2/token
```

Finally we decided to debug the `verifyIdToken` at `Google_Client` call today and found that we're getting `__PHP_Incomplete_Class_Name` from [`unserialize`](https://github.com/google/google-api-php-client/blob/master/src/Google/Cache/File.php#L59) while trying to read key `https://www.googleapis.com/oauth2/v1/certs`, contents found where:

```
O:19:"Google_Http_Request":14:{s:33:"IME-Version";s:3:"1.0";}s:14:"}s:16:"e";s:3:"200";s:18:"ication/json; charset=UTF-8";s:22:"x-content-type-options";s:7:"nosniff";s:15:"x-frame-options";s:10:"SAMEORIGIN";s:16:"x-xss-protection";s:13:"1; mode=block";s:6:"server";
s:3:"GSE";s:13:"cache-control";s:52:"public, max-age=21230, must-revalidate, no-transform";s:3:"age";s:4:"1296";s:18:"alternate-protocol";s:8:"443:quic";}s:15:"y";s:1725:"{
 "fad3a5bc6791b2977ca5fd570c11ccdc1e2c5382": "-----BEGIN CERTIFICATE-----\nMIICITCCAYqgAwIBAgIIBqosZLd67/swDQYJKoZIhvcNAQEFBQAwNjE0MDIGA1UE\nAxMrZmVkZXJhdGVkLXNpZ25vbi5zeXN
0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTAe\nFw0xNDAyMTkxMDU4MzRaFw0xNDAyMjAyMzU4MzRaMDYxNDAyBgNVBAMTK2ZlZGVy\nYXRlZC1zaWdub24uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wgZ8wDQYJKoZI\nhvcNA
QEBBQADgY0AMIGJAoGBALu5lw4FC6Q4NLA3odF/xTDsaHoDgldIccjHJGnX\nAuFy+x2mTNBMJ3AXdyragESefD5oMh2moZ5BwQ46vlztGFlP9yUi0BuJ90maY0kH\nYL8+CXOd3F8ctuKgq7eTksGt7+jk1c8bmR3Ni8HBBavU+
gOXai/akG5T5W9BZqqV\nlP0DAgMBAAGjODA2MAwGA1UdEwEB/wQCMAAwDgYDVR0PAQH/BAQDAgeAMBYGA1Ud\nJQEB/wQMMAoGCCsGAQUFBwMCMA0GCSqGSIb3DQEBBQUAA4GBAB99BRII+ZCQX1vm\nYzvP+86ETSysrEO5PoY
5iRezzEZWJI6sQhWf07Utb+ta/6FYxEDDbKeOegi5RUFj\np4e5dJM/7X1KidPujAgCCYmOegtgjfGIrkJjAsCVC/ciwCr/ILXLt4f09ao3a00T\nS/unfqKJz7QR3veDltJtazlaQjAu\n-----END CERTIFICATE-----\n",
 "b549a8c2e2c5b87eb6886114a9f8f6abfb91f98d": "-----BEGIN CERTIFICATE-----\nMIICITCCAYqgAwIBAgIIFaQcXGASEbIwDQYJKoZIhvcNAQEFBQAwNjE0MDIGA1UE\nAxMrZmVkZXJhdGVkLXNpZ25vbi5zeXN
0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTAe\nFw0xNDAyMjAxMDQzMzRaFw0xNDAyMjEyMzQzMzRaMDYxNDAyBgNVBAMTK2ZlZGVy\nYXRlZC1zaWdub24uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wgZ8wDQYJKoZI\nhvcNA
QEBBQADgY0AMIGJAoGBAN3tA/2p1+6D8rvjrcNiWZmMXRXCp78uC3uOJC9D\nohLOVQbQB60lUCecqSOUmX3+/iM+1WzV62s0ihNOgyGDXNwfibe1RBTjq5Lz5GTb\nkpvBV1+kefFKwB5y4qm0VCKvmDBNj0JGkrv0mlzNMMeKI
EEtV2G0RCSTAIfSxMD1\ncWIpAgMBAAGjODA2MAwGA1UdEwEB/wQCMAAwDgYDVR0PAQH/BAQDAgeAMBYGA1Ud\nJQEB/wQMMAoGCCsGAQUFBwMCMA0GCSqGSIb3DQEBBQUAA4GBAJPPB+QRiIuFvknZ\nl5R7HGjiAGLTQFCEtzp
35wTjFXoKWD9NlGf1IaZrUNCK/Oi6ex0HTtbrO8+fbzZo\nLvSCSCQS0IR5+3GthhoCj/fVwGg3Lc07ptOX9rC/YM/WloDCDWsFHZq6bmysxaZF\nKizFeXww12U8wJp/KrR8MNjrPY7U\n-----END CERTIFICATE-----\n"
}
";s:16:"
```

We have the option of disabling the cache but for now @davidchan0710 have patched it with, until a proper patch is released:

``` patch
From: David Chan <David Chan david@telly.com>
Date: Thu, 20 Feb 2014 16:28:25 -0800
Subject: [PATCH] Invalid Google client library's cache is return invalid value

---
 .../google-api-php-client/src/Google/Cache/File.php          | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/libs/libraries/google-api-php-client/src/Google/Cache/File.php b/libs/libraries/google-api-php-client/src/Google/Cache/File.php
index 530af80..afa1054 100644
--- a/libs/libraries/google-api-php-client/src/Google/Cache/File.php
+++ b/libs/libraries/google-api-php-client/src/Google/Cache/File.php
@@ -60,6 +60,18 @@ public function get($key, $expiration = false)
       $this->unlock($storageFile);
     }

+    // Cache data was corrupted sometimes producing __PHP_Incomplete_Class_Name.
+    // If unserialize returns bad value, delete the value and return false.
+    if (!is_bool($data) &&
+        !is_int($data) &&
+        !is_float($data) &&
+        !is_string($data) &&
+        !is_array($data) &&
+        !is_object($data)) {
+        $this->delete($key);
+        return false;
+    }
+
     return $data;
   }

-- 
1.8.5.1
```
